dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -635,6 +635,72 @@ static bool allowEarlyConstantFolding(const ExpressionActions & actions, const C
     return true;
 }
 
+
+void InterpreterSelectQuery::AnalysisResult::finalize(const ExpressionActionsChain & chain, const Context & context_, size_t where_step_num)
+{
+    if (hasPrewhere())
+    {
+        const ExpressionActionsChain::Step & step = chain.steps.at(0);
+        prewhere_info->remove_prewhere_column = step.can_remove_required_output.at(0);
+
+        Names columns_to_remove;
+        for (size_t i = 1; i < step.required_output.size(); ++i)
+        {
+            if (step.can_remove_required_output[i])
+                columns_to_remove.push_back(step.required_output[i]);
+        }
+
+        if (!columns_to_remove.empty())
+        {
+            auto columns = prewhere_info->prewhere_actions->getSampleBlock().getNamesAndTypesList();
+            ExpressionActionsPtr actions = std::make_shared<ExpressionActions>(columns, context_);
+            for (const auto & column : columns_to_remove)
+                actions->add(ExpressionAction::removeColumn(column));
+
+            prewhere_info->remove_columns_actions = std::move(actions);
+        }
+
+        columns_to_remove_after_prewhere = std::move(columns_to_remove);
+    }
+    else if (hasFilter())
+    {
+        /// Can't have prewhere and filter set simultaneously
+        filter_info->do_remove_column = chain.steps.at(0).can_remove_required_output.at(0);
+    }
+    if (hasWhere())
+        remove_where_filter = chain.steps.at(where_step_num).can_remove_required_output.at(0);
+}
+
+void InterpreterSelectQuery::AnalysisResult::removeExtraColumns()
+{
+    if (hasFilter())
+        filter_info->actions->prependProjectInput();
+    if (hasWhere())
+        before_where->prependProjectInput();
+    if (hasHaving())
+        before_having->prependProjectInput();
+}
+
+void InterpreterSelectQuery::AnalysisResult::checkActions()
+{
+    /// Check that PREWHERE doesn't contain unusual actions. Unusual actions are that can change number of rows.
+    if (hasPrewhere())
+    {
+        auto check_actions = [](const ExpressionActionsPtr & actions)
+        {
+            if (actions)
+                for (const auto & action : actions->getActions())
+                    if (action.type == ExpressionAction::Type::JOIN || action.type == ExpressionAction::Type::ARRAY_JOIN)
```

Comment:
Isn't it better to define canChangeRowsNumber helper in ExpressionActions?

Reply:
It's a refactoring not a rewriting. Do not change any code logic.